### ISSUE TITLE
libdeflate_gzip_compress fails for 32-bit windows builds

### ISSUE THREAD

**Issue by dosh1974** [ebiggers/libdeflate#35]
Getting an exception when using the pre-built windows dll  [1.0] (called from C#).
64 bit gzip works.
32 bit and 64 bit deflate works.
My implementation is identical for gzip and deflate (with exception of the function calls to the DLL).

"System.AccessViolationException [0x80004003] Attempted to read or write protected memory. This is often an indication that other memory is corrupt."

Tested with a bunch of data files from the Silesia corpus ("dickens", "xml", "mozilla" etc).

**Comment by ebiggers** [comment]
Are you using the cdecl calling convention?  See this section of the README file:

> Windows developers: note that the calling convention of libdeflate.dll is cdecl.
> This differs from the Windows API, which uses stdcall.  If you call into
> libdeflate.dll using a non-C/C++ language, or dynamically using LoadLibrary(),
> make sure to use the cdecl calling convention.  Using the wrong calling
> convention may crash your application.

**Comment by dosh1974** [comment]
After testing some more, I'm pretty sure that it's reading past (or before) the input buffer.
If I allocate a bigger data segment and compresses from that it works fine.,.
If I had the C++ version up and running I would have allocated 3 continuous memory pages (VirtualAlloc as readonly). Marked page 0 and 2 as fully protected (VirtualProtect), then compress page 1 and see what happens.
But I don't have any more time to spend on this issue and again since X86 isn't my concern I'll leave it at this... I hope you find the cause :)

**Comment by ebiggers** [comment]
@dosh1974 any update?  If not I will close this, since as far as I can tell this isn't a libdeflate problem...

**Comment by dosh1974** [comment]
I didn't dig much deeper, but I'm still no 100% convinced that there isn't something fishy going on (like reading past input buffer in some cases, like odd input size or so).
But, I'm not using 32 bit so I don't really care so please close if you wish.

**Comment by dosh1974** [comment]
Well now you had me write a repro case..
I've attached a Visual Studio 2017 project with a case showing the issue.
[LibDeflateTest.zip](https://github.com/ebiggers/libdeflate/files/2652898/LibDeflateTest.zip)

It seems like it's data dependent, so my guess is it depends on what part of the compressor reads the final bytes.

**Comment by dosh1974** [comment]
Including the and using the latest source (instead of using the pre-built binaries for windows), seems to work, leaving a couple of options.
- Something has been fixed (changed) since the v1.0 release.
- The compiler used for building the v1.0 release binaries is doing some bad codegen (optimizations?).

I don't have any compiler but VS installed atm and I guess you're using some very specific, fine tuned,  compiler setup to maximize performance, so I let it rest for now :)

**Comment by dosh1974** [comment]
Here's the error in VS.
The address is tries to read is -1, in __libdeflate_aligned_free, this suggests that my initial assupmtion of reading before / bast the input buffer is wrong.
See screenshot:
![vserror](https://user-images.githubusercontent.com/5424329/49634953-7c774200-f9fe-11e8-8e63-a51aa92908b7.png)


Further I noticed that sometimes it does work!
That suggests to me that there might be an uninitialized memory/variable problem since I don't expect a compressor to use any randomness in it's process and I haven't seen any indication of it being multi-threaded (should be 100% deterministic IMO).
See:
![success](https://user-images.githubusercontent.com/5424329/49635050-c6602800-f9fe-11e8-8899-8f321b48a099.png)

It makes sense that the issue persist in both the statically linked (C++) and the DLL (my initial issue, from C#) if it's the same compiler.
Odd that I can't reproduce when I build everything using VS and you can't reproduce when building everything using MinGW.
This suggests that there are some interfacing issues, but since the deflate works 100% using identical signatures etc it doesn't seem very likely.

I've also attached (debug / release) binaries.
[LibDeflateTestBinaries.zip](https://github.com/ebiggers/libdeflate/files/2656406/LibDeflateTestBinaries.zip)

**Comment by dosh1974** [comment]
As for my CPU, it's a modest I5:

![cpu](https://user-images.githubusercontent.com/5424329/49637084-ac294880-fa04-11e8-84d2-94d3bb3ef708.png)

**Comment by dosh1974** [comment]
Figured out why building from source on VS works, it not using any of the SSE / AVX optimized versions... made a half arsed attempt of getting it working in VS but gave up... but at least that's probably why it worked.

**Comment by jibsen** [comment]
Digging around a bit, I think this may be caused by an unaligned store to a temporary in `crc32_pclmul_avx()` (makes sense, since computing crc32 is pretty much the only difference between gzip and deflate). Visual Studio misinterprets the debug info and claims the error is in `libdeflate_aligned_free()`, but the example never calls the free function.
~~~
01320D40  lea         ebx,[edx+40h]  
01320D43  vmovdqa     xmm2,xmmword ptr [edx+10h]  
01320D48  cmp         edi,ebx  
01320D4A  vmovdqa     xmm4,xmmword ptr [edx+20h]  
01320D4F  vmovdqa     xmm5,xmmword ptr [edx+30h]  
01320D54  je          __libdeflate_aligned_free+32Dh (01320DCDh)  
01320D56  mov         eax,ebx  
01320D58  vmovdqa     xmm6,xmmword ptr ds:[13BAC20h]  
01320D60  vpclmulqdq  xmm1,xmm4,xmm6,0  
01320D66  add         eax,40h  
01320D69  vpclmulqdq  xmm4,xmm4,xmm6,11h  
01320D6F  vpclmulqdq  xmm7,xmm0,xmm6,0  
01320D75  vmovaps     xmmword ptr [esp],xmm1 ; <-- esp is 00EFF8DC
~~~

The problem might be that GCC assumes the stack is 16-byte aligned on entry, but x86 MSVC does not. So when you call the function in libdeflate.lib (which is compiled with GCC 7.3.0) from x86 MSVC this can happen.

**Comment by jibsen** [comment]
A little further investigation suggests the problem is that GCC does not add stack alignment code when you use the target attribute alone. If you compile with the proper flags, or add the [`force_align_arg_pointer`](https://gcc.gnu.org/onlinedocs/gcc/x86-Function-Attributes.html) attribute, it does.

Small example:
~~~.c
#include <wmmintrin.h>

__attribute__((target("sse2")))
int foo(__m128i a)
{
	__m128i n = _mm_srli_si128(a, 8);
	
	return *(unsigned char *) &n;
}
~~~

compiles with `gcc -O2 -m32 -S -c foo.c` to
~~~.asm
_foo:
	movdqa	%xmm0, %xmm1
	subl	$44, %esp
	psrldq	$8, %xmm1
	movaps	%xmm1, (%esp)
	movzbl	(%esp), %eax
	addl	$44, %esp
	ret
~~~
which will crash if the stack is not aligned on entry. Whereas if you add the `force_align_arg_pointer` attribute (or compile with `-msse2` or `-mincoming-stack-boundary=2`), you get
~~~.asm
_foo:
	pushl	%ebp
	movdqa	%xmm0, %xmm1
	psrldq	$8, %xmm1
	movl	%esp, %ebp
	andl	$-16, %esp
	subl	$32, %esp
	movaps	%xmm1, (%esp)
	movzbl	(%esp), %eax
	leave
	ret
~~~

**Comment by ebiggers** [comment]
Thanks for the extra information @dosh1974, and thanks for the detailed analysis @jibsen!  It seems that's exactly what the problem was.  gcc assumes the stack is 16-byte aligned and can take advantage of that to generate aligned SSE or AVX load/stores to the stack.  But actually MSVC only guarantees 4-byte alignment.  I've fixed this by adding `__attribute__((force_align_arg_pointer))` to libdeflate's API entry points when it's compiled for 32-bit Windows with gcc.

I published a prerelease build of the fixed Windows binaries to https://github.com/ebiggers/libdeflate/releases/tag/v1.0-8-gd6d50c6.  Please verify that they work for you now -- thanks!

**Comment by ebiggers** [comment]
The fix is in the v1.1 release now.
